fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementInUnionRule.java

```diff
@@ -0,0 +1,262 @@
+import com.apple.foundationdb.record.query.plan.temp.properties.OrderingProperty;
+import com.apple.foundationdb.record.query.predicates.LiteralValue;
+import com.apple.foundationdb.record.query.predicates.QuantifiedColumnValue;
+import com.apple.foundationdb.record.query.predicates.Value;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.SetMultimap;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.apple.foundationdb.record.query.plan.temp.matchers.MultiMatcher.some;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.QuantifierMatchers.forEachQuantifier;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.QuantifierMatchers.forEachQuantifierOverPlans;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.RelationalExpressionMatchers.explodeExpression;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.RelationalExpressionMatchers.selectExpression;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.SetMatcher.exactlyInAnyOrder;
+
+/**
+ * A rule that implements a logical filter around a {@link RecordQueryPlan} as a {@link RecordQueryFilterPlan}.
+ */
+@API(API.Status.EXPERIMENTAL)
+@SuppressWarnings("PMD.TooManyStaticImports")
+public class ImplementInUnionRule extends PlannerRule<SelectExpression> {
+    private static final CollectionMatcher<RecordQueryPlan> innerPlansMatcher = some(RecordQueryPlanMatchers.anyPlan());
+    private static final BindingMatcher<Quantifier.ForEach> innerQuantifierMatcher = forEachQuantifierOverPlans(innerPlansMatcher);
+    private static final BindingMatcher<ExplodeExpression> explodeExpressionMatcher = explodeExpression();
+    private static final BindingMatcher<Quantifier.ForEach> explodeQuantifierMatcher = forEachQuantifier(explodeExpressionMatcher);
+
+    private static final BindingMatcher<SelectExpression> root =
+            selectExpression(exactlyInAnyOrder(innerQuantifierMatcher, explodeQuantifierMatcher));
+
+    public ImplementInUnionRule() {
+        super(root, ImmutableSet.of(OrderingAttribute.ORDERING));
+    }
+
+    @SuppressWarnings({"unchecked", "java:S135"})
+    @Override
+    public void onMatch(@Nonnull PlannerRuleCall call) {
+        final PlanContext context = call.getContext();
+
+        final PlannerBindings bindings = call.getBindings();
+
+        final Optional<Set<Ordering>> requiredOrderingsOptional = call.getInterestingProperty(OrderingAttribute.ORDERING);
+        if (!requiredOrderingsOptional.isPresent()) {
+            return;
```

Comment:
It's possible that there is no required ordering (yet) that has been pushed down sufficiently. Such an ordering can include an ordering that preserves the existing ordering in the absence of an explicitly given ordering. The guaranteed from the planner is that once a required ordering becomes available the rule is reevaluated (that is declared in the constructor of the rule). So we can safely say here that we won't do a thing here if such a required ordering does not exist. Maybe this can be done as a matcher o avoid confusion.